codegen-sbt/src/main/scala/caliban/codegen/CompileTimeCalibanPlugin.scala

```diff
@@ -0,0 +1,450 @@
+
+    // ## Plugin task
+    lazy val ctCalibanClientGenerate: TaskKey[Seq[File]] = taskKey[Seq[File]](
+      "Generate Caliban Client(s) code at compile time. Automatically configured to be triggered when compilation is triggered"
+    )
+  }
+  import autoImport._
+
+  /**
+   * I have to apologize for the readability and complexity of this code.
+   * I did my best with the limited knowledge I have and the constraints that sbt is putting on us 😕
+   */
+  private lazy val pluginSettings =
+    inTask(ctCalibanClient)(
+      Seq(
+        ctCalibanClientsSettings := Seq.empty,
+        ctCalibanClientGenerate := {
+          // That helped: https://stackoverflow.com/q/26244115/2431728
+          Def.taskDyn {
+            import Functions._
+
+            val log = streams.value.log("ctCalibanClient")
+
+            val clientsSettings: Seq[Project] = (ctCalibanClient / ctCalibanClientsSettings).value
+            if (clientsSettings.isEmpty) Def.task { log.error(helpMsg); Seq.empty[File] }
+            else {
+              val baseDirValue: String = (thisProject / baseDirectory).value.absolutePath
+
+              def generateSources: Def.Initialize[Task[Seq[File]]] =
+                Def.taskDyn {
+                  log.info(s"ctCalibanClient - Starting to generate...")
+
+                  Def.task {
+                    clientsSettings
+                      .flatTraverseT[File] { serverProject: Project =>
+                        Def.taskDyn {
+                          ensureCompiled(serverProject).value
+
+                          val serverMetadata = {
+                            val serverTargetDir = (serverProject / target).value.getAbsolutePath
+                            file(s"$serverTargetDir/ctCalibanServer/metadata")
+                          }
+
+                          if (!serverMetadata.exists()) Def.task(Seq.empty[File])
+                          else {
+                            val generatedRefs: Seq[(File, String, String)] =
+                              sbt.IO.readLines(serverMetadata, defaultCharset).map { line =>
+                                val Array(generatorFile, generatorRef, packageName) = line.split("#")
+                                (file(generatorFile), generatorRef, packageName)
+                              }
+
+                            generatedRefs
+                              .flatTraverseT[File] { case (generatorFile, generatorRef, packageName) =>
+                                Def.taskDyn {
+                                  def listGeneratedClientsFiles: Set[File] = {
+                                    val toPathDir: File = file(Utils.toPathDir(baseDirValue, packageName))
+
+                                    if (!toPathDir.exists()) {
+                                      sbt.IO.createDirectory(toPathDir)
+                                      fsync(toPathDir.toPath)
+
+                                      log.warn(s"toPathDir - exists: ${toPathDir.exists()}")
```

Comment:
These could be removed, or at least changed to debug.

Reply:
I'm trying to debug the error I'm experiencing with sbt: https://github.com/sbt/sbt/issues/6655